hibernate-core/src/main/java/org/hibernate/query/criteria/internal/expression/MapEntryExpression.java

```diff
@@ -25,41 +27,41 @@
  */
 public class MapEntryExpression<K,V>
 		extends ExpressionImpl<Map.Entry<K,V>>
-		implements Expression<Map.Entry<K,V>>, Serializable {
+		implements CompoundSelection<Map.Entry<K,V>>, Serializable {
 
-	private final PathImplementor origin;
-	private final MapAttribute<?, K, V> attribute;
+	private final MapAttributeJoin<?, K, V> original;
 
 	public MapEntryExpression(
 			CriteriaBuilderImpl criteriaBuilder,
 			Class<Map.Entry<K, V>> javaType,
-			PathImplementor origin,
-			MapAttribute<?, K, V> attribute) {
-		super( criteriaBuilder, javaType);
-		this.origin = origin;
-		this.attribute = attribute;
-	}
-
-	public MapAttribute<?, K, V> getAttribute() {
-		return attribute;
+			MapAttributeJoin<?, K, V> original) {
+		super( criteriaBuilder, javaType );
+		this.original = original;
 	}
 
+	@Override
 	public void registerParameters(ParameterRegistry registry) {
 		// none to register
 	}
 
+	@Override
 	public String render(RenderingContext renderingContext) {
 		if ( renderingContext.getClauseStack().getCurrent() == Clause.SELECT ) {
-			return "entry(" + path( renderingContext ) + ")";
+			return "entry(" + original.render( renderingContext ) + ")";
 		}
 
 		// don't think this is valid outside of select clause...
 		throw new IllegalStateException( "illegal reference to map entry outside of select clause." );
 	}
 
-	private String path(RenderingContext renderingContext) {
-		return origin.getPathIdentifier()
-				+ '.'
-				+ ( (Renderable) getAttribute() ).render( renderingContext );
+	@Override
+	public boolean isCompoundSelection() {
+		return true;
 	}
+
+	@Override
+	public List<Selection<?>> getCompoundSelectionItems() {
+		return Arrays.asList( original.key(), original.value() );
```

Comment:
Force pushed to add an `@Override` here.

I remember now why I didn't implement it as a `CompoundSelection`, I didn't know how to implement this method.

It certainly looks better this way.